controller/monitor/snapshot_monitor.go

```diff
@@ -0,0 +1,588 @@
+	}
+
+	if dataIntegrityCronJob == m.existingDataIntegrityCronJob && m.checkScheduler.Len() > 0 {
+		return nil
+	}
+
+	if m.checkScheduler.Len() > 0 {
+		m.checkScheduler.Remove(m.checkSnapshots)
+	}
+
+	job, err := m.checkScheduler.Cron(dataIntegrityCronJob).Do(m.checkSnapshots)
+	if err != nil {
+		return errors.Wrap(err, "failed to schedule snapshot check job")
+	}
+
+	m.Lock()
+	m.existingDataIntegrityCronJob = dataIntegrityCronJob
+	m.Unlock()
+
+	m.checkScheduler.StartAsync()
+
+	m.logger.WithField("monitor", monitorName).Infof("Next snapshot check job will be executed at %v", job.NextRun())
+
+	return nil
+}
+
+func (m *SnapshotMonitor) GetData() (interface{}, error) {
+	m.RLock()
+	defer m.RUnlock()
+	return m.SnapshotMonitorStatus, nil
+}
+
+func (m *SnapshotMonitor) addToInProgressSnapshotCheckTasks(snapshotName string) bool {
+	m.inProgressSnapshotCheckTasksLock.Lock()
+	defer m.inProgressSnapshotCheckTasksLock.Unlock()
+
+	_, ok := m.inProgressSnapshotCheckTasks[snapshotName]
+	if ok {
+		m.logger.WithField("monitor", monitorName).Debugf("snapshot %s is being checked", snapshotName)
+		return false
+	}
+	m.inProgressSnapshotCheckTasks[snapshotName] = struct{}{}
+
+	return true
+}
+
+func (m *SnapshotMonitor) deleteFromInProgressSnapshotCheckTasks(snapshotName string) {
+	m.inProgressSnapshotCheckTasksLock.Lock()
+	defer m.inProgressSnapshotCheckTasksLock.Unlock()
+
+	delete(m.inProgressSnapshotCheckTasks, snapshotName)
+}
+
+func (m *SnapshotMonitor) RunTask(arg interface{}) error {
+	task, ok := arg.(snapshotCheckTask)
+	if !ok {
+		return fmt.Errorf("failed to assert value: %v", arg)
+	}
+
+	if !m.addToInProgressSnapshotCheckTasks(task.snapshotName) {
+		return nil
+	}
+	defer m.deleteFromInProgressSnapshotCheckTasks(task.snapshotName)
+
+	engine, err := m.ds.GetVolumeCurrentEngine(task.volumeName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get engine for volume %v", task.volumeName)
+	}
+
+	if err := m.checkVolumeNotInMigration(engine.Spec.VolumeName); err != nil {
+		return err
+	}
+
+	engineCliClient, err := m.getEngineBinaryClient(engine.Spec.VolumeName)
+	if err != nil {
+		return err
+	}
+
+	engineClientProxy, err := engineapi.GetCompatibleClient(engine, engineCliClient, m.ds, m.logger, m.proxyConnCounter)
+	if err != nil {
+		return err
+	}
+	defer engineClientProxy.Close()
+
+	m.Lock()
+	m.SnapshotCheckState = longhorn.SnapshotCheckStateInProgress
+	m.Unlock()
+
+	defer func() {
+		m.Lock()
+		defer m.Unlock()
+		m.SnapshotCheckState = longhorn.SnapshotCheckStateIdle
+	}()
+
+	err = m.requestSnapshotHashing(engine, engineClientProxy, task.snapshotName, task.changeEvent)
+	if err != nil {
+		return err
+	}
+
+	return m.waitForSnapshotHashingComplete(engine, engineClientProxy, task.snapshotName)
+}
+
+func (m *SnapshotMonitor) requestSnapshotHashing(engine *longhorn.Engine, engineClientProxy engineapi.EngineClientProxy,
+	snapshotName string, changeEvent bool) error {
+	rehash := false
+	if !changeEvent {
```

Comment:
Yeah. Because the one snapshot CR might be updated many times in a short period. I expect it just calculates checksum once. If `refresh` is false and the checksum is existing, don't need to calculate it again if the `ctime` is not changed.

The periodc snapshot check mechanism will do the regular checks.

Reply:
OK. In other words, the full hash will be issued by the cron job only, no matter if the field is `enabled` or `fast-check`.

Reply:
Yes. You're right.

Reply:
OK. Please leave a comment here.

Reply:
Added. Thank you.